filestore/util.go

```diff
@@ -0,0 +1,184 @@
+	StatusFileNotFound Status = 11 // Backing File Not Found
+	StatusFileChanged  Status = 12 // Contents of the file changed
+	StatusOtherError   Status = 20 // Internal Error, likely corrupt entry
+	StatusKeyNotFound  Status = 30
+)
+
+func (s Status) String() string {
+	switch s {
+	case StatusOk:
+		return "ok"
+	case StatusFileError:
+		return "error"
+	case StatusFileNotFound:
+		return "no-file"
+	case StatusFileChanged:
+		return "changed"
+	case StatusOtherError:
+		return "ERROR"
+	case StatusKeyNotFound:
+		return "missing"
+	default:
+		return "???"
+	}
+}
+
+func (s Status) Format() string {
+	return fmt.Sprintf("%-7s", s.String())
+}
+
+type ListRes struct {
+	Status   Status
+	ErrorMsg string
+	Key      *cid.Cid
+	FilePath string
+	Offset   uint64
+	Size     uint64
+}
+
+func (r *ListRes) FormatLong() string {
+	switch {
+	case r.Key == nil:
+		return "?????????????????????????????????????????????????"
+	case r.FilePath == "":
+		return r.Key.String()
+	default:
+		return fmt.Sprintf("%-50s %6d %s %d", r.Key, r.Size, r.FilePath, r.Offset)
+	}
+}
+
+func List(fs *Filestore, key *cid.Cid) *ListRes {
+	return list(fs, false, key)
+}
+
+func ListAll(fs *Filestore) (func() *ListRes, error) {
+	return listAll(fs, false)
+}
+
+func Verify(fs *Filestore, key *cid.Cid) *ListRes {
+	return list(fs, true, key)
+}
+
+func VerifyAll(fs *Filestore) (func() *ListRes, error) {
+	return listAll(fs, true)
+}
+
+func list(fs *Filestore, verify bool, key *cid.Cid) *ListRes {
+	dobj, err := fs.fm.getDataObj(key)
+	if err != nil {
+		return mkListRes(key, nil, err)
+	}
+	if verify {
+		_, err = fs.fm.readDataObj(key, dobj)
+	}
+	return mkListRes(key, dobj, err)
+}
+
+func listAll(fs *Filestore, verify bool) (func() *ListRes, error) {
+	q := dsq.Query{}
+	qr, err := fs.fm.ds.Query(q)
+	if err != nil {
+		return nil, err
+	}
+
+	return func() *ListRes {
+		cid, dobj, err := next(qr)
+		if dobj == nil && err == nil {
+			return nil
+		} else if err == nil && verify {
+			_, err = fs.fm.readDataObj(cid, dobj)
+		}
+		return mkListRes(cid, dobj, err)
+	}, nil
+}
+
+func next(qr dsq.Results) (*cid.Cid, *pb.DataObj, error) {
+	v, ok := qr.NextSync()
+	if !ok {
+		return nil, nil, nil
+	}
+
+	k := ds.RawKey(v.Key)
+	c, err := dshelp.DsKeyToCid(k)
+	if err != nil {
+		return nil, nil, fmt.Errorf("decoding cid from filestore: %s", err)
+	}
+
+	data, ok := v.Value.([]byte)
+	if !ok {
+		return c, nil, fmt.Errorf("stored filestore dataobj was not a []byte")
+	}
+
+	var dobj pb.DataObj
+	if err := proto.Unmarshal(data, &dobj); err != nil {
+		return c, nil, err
+	}
+
+	return c, &dobj, nil
+}
+
+func mkListRes(c *cid.Cid, d *pb.DataObj, err error) *ListRes {
+	status := StatusOk
+	errorMsg := ""
+	if err != nil {
+		if err == ds.ErrNotFound || err == blockstore.ErrNotFound {
+			status = StatusKeyNotFound
+		} else if err, ok := err.(*CorruptReferenceError); ok {
+			switch err.Code {
+			case FileError:
+				status = StatusFileError
```

Comment:
The set of codes returned in ListRes is larger than the codes for CorruptReferenceError.  For a little less type safety I can have CorruptReferenceError use the same codes in ListRes if you prefer it that way,